### PR TITLE
Add `npartitions` and `divisions`-based repartitioning

### PR DESCRIPTION
**Pull request by rjzamora**
Adds an abstract `Repartition` expression, and subclasses for `npartitions` and `divisions`-based repartitioning.

### CODE REVIEW

File: dask_match/repartition.py [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+
+    def _divisions(self):
+        if self.n is not None:
+            return self.simplify()._divisions()
+        return self.new_divisions
+
+    def simplify(self):
+        if self.n is not None:
+            if self.n < self.frame.npartitions:
+                return ReducePartitionCount(self.frame, self.n)
```
Comment (mrocklin) [inline dask_match/repartition.py:35]: I suggest that all of these classes should maybe start with `Repartition`, like `RepartitionToMorePartitions` and `RepartitionToFewerPartitions`.  Mostly I care about Repartition being in the name, and probably the first term.

File: dask_match/repartition.py [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+                isinstance(frame, cls)
+                and not isinstance(other, Expr)
+            )
+        )
+        for op in [Filter, Projection]:
+            pattern = Pattern(op(frame, other), supported)
+            yield ReplacementRule(pattern, functools.partial(reorder, op=op))
+
+
+class RepartitionImpl(Repartition):
```
Comment (mrocklin) [inline dask_match/repartition.py:102]: I suggest maybe the following names

```python
Repartition  # the super-class
    RepartitionAbstract
    RepartitionToMore
    RepartitionToFewer
    RepartitionDivisions
```

**Review by mrocklin — COMMENTED** [review]
Seems fine to me.  Thanks @rjzamora .  A small comment about names.

File: dask_match/repartition.py (hunk 1 above)
Comment (rjzamora) [reply dask_match/repartition.py:35]: Cool - Make sense to me.

File: dask_match/repartition.py [hunk 3]
```diff
@@ -1,22 +1,113 @@
...
+        supported = CustomConstraint(
+            lambda frame, other: (
+                # filter-pushdown gets messy if other is an Expr
+                isinstance(frame, cls)
+                and not isinstance(other, Expr)
+            )
+        )
+        for op in [Filter, Projection]:
+            pattern = Pattern(op(frame, other), supported)
+            yield ReplacementRule(pattern, functools.partial(reorder, op=op))
```
Comment (rjzamora) [inline dask_match/repartition.py:103]: Will need to rewrite this with `simplify_up` - So, I'd say we should treat #50 as a dependency for this PR.

Comment (rjzamora) [reply dask_match/repartition.py]: Done.

File: dask_match/repartition.py [hunk 4]
```diff
@@ -1,22 +1,101 @@
...
 
+    def _divisions(self):
+        if self.n is not None:
+            return self.simplify()._divisions()
+        return self.new_divisions
+
+    def _simplify_down(self):
+        if type(self) != Repartition:
+            # This simplify logic should not be inherited
+            return None
```
Comment (mrocklin) [inline dask_match/repartition.py:35]: Interesting pattern

File: dask_match/repartition.py [hunk 5]
```diff
@@ -1,22 +1,101 @@
...
+        elif self.new_divisions:
+            return RepartitionDivisions(self.frame, self.new_divisions, self.force)
+        else:
+            raise NotImplementedError()
+
+    def _simplify_up(self, parent):
+        # Move filter or column projection before repartitioning
+        if isinstance(parent, (Filter, Projection)) and not isinstance(
+            parent.operands[1], Expr
+        ):
```
Comment (mrocklin) [inline dask_match/repartition.py:89]: Isn't the predicate of `Filter` likely to be an expression?

I wonder if it makes sense to break these two cases out and repeat a little.  Some duplication, but maybe the code is easier to understand.  Happy either way.

File: dask_match/repartition.py [hunk 6]
```diff
@@ -1,22 +1,101 @@
...
+            op = type(parent)(self.frame, *parent.operands[1:])
+            return type(self)(op, *self.operands[1:])
```
Comment (mrocklin) [inline dask_match/repartition.py:91]:
```python
if isinstance(parent, Projection):
    return self.frame[parent.columns].repartition(*self.operands[1:])
if isinstance(parent, Filter):
    ...
```

Comment (mrocklin) [reply dask_match/repartition.py:91]: Or I guess we want to use the proper Op, given that we may already have lowered.  In practice though I expect `simplify_up` to be called before `simplify_down`.  It's best not to rely on that though.

**Review by mrocklin — APPROVED** [review]

File: dask_match/repartition.py (hunk 5 above)
Comment (rjzamora) [reply dask_match/repartition.py]: >Isn't the predicate of Filter likely to be an expression?

Ah right - I confused myself here.

Comment (mrocklin) [reply dask_match/repartition.py]: A common effect of DRY code 😉

**Pull request merged**